### ISSUE TITLE
Support vscode's "Remote Development extensions"

### ISSUE THREAD

**Issue by rachellawson** [ddev/ddev#1569]
**Is your feature request related to a problem? Please describe.**
The Visual Studio Code team recently blogged that they are building a "VSCode server component" that means there will be *no need to have filesystem sharing between host and container on hosts running Mac OS and Windows*. 

See https://code.visualstudio.com/docs/remote/containers

**Describe the solution you'd like**
I'd like us to work towards an optional container that can be configured that holds the server-side components mentioned in the article above and allows access to work on the code via vscode without needing to set up filesystem sharing via nfs, syncing etc.

**Additional context**
Will be trying to ge this work work myself but also want to gather interested parties in one place so we can help each other!

**Comment by rfay** [comment]
I'm not sure you need anything from ddev to do this. The post-start hook allows installing extra packages as in https://ddev.readthedocs.io/en/stable/users/extending-commands/#adding-additional-php-modules-example

However, in next release of ddev (next week) there will be a much easier  and faster way to install extra packages.

**Comment by rachellawson** [comment]
It was mostly about configuring the container host ports and writing an appropriate devcontainer.json file that I was thinking. I thought this would be good in `dev config` and then viscode would automatically connect correctly without the developer having to do much. It's a little more than just installing a package (but not much)

**Comment by rachellawson** [comment]
and, if there is going to be a new way to install extra packages, I'll wait for that before starting!

**Comment by rfay** [comment]
You could easily get the packages in there and maybe you need a docker-compose.vscode.yaml? https://ddev.readthedocs.io/en/stable/users/extend/customization-extendibility/

**Comment by joshmiller83** [comment]
If this truly would remove the bottleneck of filesystem syncrhonization, I would consider moving from phpstorm to vs code and from lando to ddev.

**Comment by rfay** [comment]
@joshmiller83 - Have you been using NFS? I assume you're on macOS. It makes  a huge difference. https://ddev.readthedocs.io/en/stable/users/performance/

As above, I'm not at all sure that anything is required here to use vscode's nice in-docker stuff. The web container is there waiting for it...

**Comment by JPustkuchen** [comment]
Just tried the same thing and I think it would improve the development workflow and sharing of project based best-practice VSCode configuration / extensions a lot.

Did anyone manage to get this up and running with VSCode attached to a ddev container with .devcontainer/devcontainer.json standardization?

For us it works with custom docker-compose.yml and Dockerfile, but not combined with DDev.

**Comment by rfay** [comment]
Maybe you could explain the full detail of what you tried, and share what you do that worked, and it could spur some investigation.

**Comment by joshsedl** [comment]
Hey @rfay,

me and @JPustkuchen, tried to attach VS-Code to our Lando-Project, with preexisting PHP and Drupal Settings/Extensions for VS-Code.

The main goal is, having a Lando-Debug-Environment with integrated VS-Code "Best Practice" PHP/Drupal programming Settings, so we don't have to search for the different settings and extensions manually and install them in our local VS-Code Environment.

So basically an easy Plug and Play programming Environment, for our Team.

I already created a "dummy-devcontainer-php" Environment for PHP programming purposes and now I tried to fork @JPustkuchen "lando-drupal8-test-debugging" repository and added an .vscode Folder, for Extensions and Settings.

It kinda works, but there is no proper Lando Support for the Remote Development Extension so we can't add it as a service. We could add it as a custom service, but then we can't attach VSCode. Only create it's own Container with .devcontainer. The problem with that is, that XDebug won't work with Lando's Version of XDebug but with it's own, defined in the Dockerfile/docker-compose.

So maybe there will be official lando support in the future who knows.

However, we haven't touched ddev yet, but will, at a given time.

The Environment @JPustkuchen reffered to: https://github.com/joshsedl/php-devcontainer-dummy

The Lando-Environment with VSCode "Best Practise" Settings: https://github.com/joshsedl/lando-drupal8-test-debugging

I hope this information can help, and I wish you happy coding :)

**Comment by joshsedl** [comment]
For further explanation, since this is a ddev issue post and not a lando issue post.
The main problem regarding ddev and lando is, that there is currently no way, to attach VS-Code to a running Container via .devcontainer definitions. But correct me if I'm wrong.

Furthermore we want to switch to ddev in the Future, so if anybody has some good sources, regarding attaching VS-Code to a running ddev Container with  predefined settings through .devcontainer definitions, feel free to share!

**Comment by rfay** [comment]
From a quick read of the [original link](https://code.visualstudio.com/docs/remote/containers) it seems that vscode wants to start the container itself. Of course, that's not going to work out, because ddev already has it running. You could use ddev's web image as a standalone web container, or run a second one and have it join in with an existing running project, but mostly it seems like a whole different concept.

**Comment by JPustkuchen** [comment]
Hi @rfay,

thank you for your quick reply! Yes you're absolutely right. And it's definitely not the fault of ddev or lando that this doesn't work as expected. We also thought about using the images like you wrote but well that looks like a further workaround.

So I guess we should treat this issue as future idea and hope for the VSCode developers to extend their concept of attaching to a container via UI to attach by container name or script in a .devcontainer file.

As @joshsedl wrote we could imagine a scenario like
a) starting up ddev and attaching to the webserver container from vscode as defined in .devcontainer
or
b) have ddev running and attach into that running container by commands in .devcontainer

The main goal, from my perspective, would be that ALL dependencies could be IN the container, including VSCode Extensions, project specific settings, launch configuration etc.
If you're developing for example in multiple projects / languages, this allows for a simple and clean switch between your separated projects.

If you have a look at the .devcontainer attributes, that's what it does, but sadly not for a more complex attach yet :( So finally I'd say the concept behind the .devcontainer file is what's interesting for ddev, but .devcontainer didn't have the focus that much on attaching yet?

**Comment by joshsedl** [comment]
@rfay and all the others interested here:

Did someone manage to attach to a running ddev container from console? I just can't find the start parameter or something similar for VSCode to do this without UI...

The idea is to have a launch script which starts the ddev container and vscode attached to the web container.
Like this (but no idea yet how to make this work):

```
#!/bin/bash
ddev start
ddev ssh
cd /var/www/html
code .  # Problem: vscode is not installed in the container. So this can't work
```
or
```
#!/bin/bash
ddev start
code 'ddev ssh -d="/var/www/html/"' # Idea is to tell code to attach to the container as starting argument like . for the current directory
```

The benefit would be to have one script to execute and you're up and running in the container with the correct environment and tools.

Any ideas? :(

**Comment by rfay** [comment]
I would expect you could run the code proxy from within the container - that's what their remote execution is for, rather than trying it from outside.

You might try installing "real ssh" and use that, https://github.com/drud/ddev-contrib/blob/master/recipes/sshd

But I don't see why you can't `ddev ssh` from within code...

**Comment by Morgy93** [comment]
I don't understand why people want to attach to containers, rather than working on the host + ddev ssh.

Anyway, there's something like 
`code --remote dev-container+<name> <path>` or
`code --remote ssh-remote+<host> <remote_path>` and even URI specs like
`vscode://vscode-remote/k8s-container+.....`

Maybe that's something you can work with @joshsedl 

Ref:
https://github.com/microsoft/vscode-remote-release/issues/656#issuecomment-584209378
https://github.com/microsoft/vscode-remote-release/issues/4684#issuecomment-871232719

**Comment by rachellawson** [comment]
One reason that would be particularly useful for connecting like this is that there would no longer need to be a need for the IDE to have access to the filesystem inside DDEV. So, no nfs etc, so speeeeeedy

**Comment by joshsedl** [comment]
Thank you everyone for your fast replies!

I just tried @Morgy93 answer but couldn't get it to work right away. Maybe a typo, maybe something else. It basically just started VS-Code in WSL and not in my container.

Anyways, I will try again tommorow with a clear mind.
Still, thanks for caring about such an old issue! I will reply again if it works in the morning :)

**Comment by joshsedl** [comment]
Ok I'm stupid, didn't know the + is an ascii path to my hexadecimal encoded devcontainer Folder. Cheers everyone!

**Comment by rfay** [comment]
I hope you'll take a moment to write your whole solution here when you're comfortable with it @joshsedl , thanks.

**Comment by JPustkuchen** [comment]
Thank you all very much for your help and fast replies! It still seems quite (officially) undocumented which parameters to use for such vscode CLI experiments...

https://stackoverflow.com/questions/60379221/how-to-attach-a-remote-container-using-vscode-command-line/63715551#63715551

This one explains how it **might** work, but still you'd need the container path in ascii representation. I guess it's the value which vscode shows last when using F1 + "Remote-Containers: Attach to a running container" in the list of available containers.

Some other helpful resources so far:
https://stackoverflow.com/search?q=%5Bvscode-remote%5D+cli
https://stackoverflow.com/questions/57382502/vscode-open-remote-from-cli

Still seems kind of ugly not to be able to attach to a running container by name...

**Comment by joshsedl** [comment]
So basically this is my solution, it's not perfect, but it works just fine:
~~~
code --folder-uri="vscode-remote://attached-container+{container-name-in-hex}%{path-in-container}"
~~~
I created a ddev environment with an appropriate devcontainer.json, see:
https://code.visualstudio.com/docs/remote/attach-container
https://code.visualstudio.com/docs/remote/devcontainerjson-reference

After i started up my ddev environment, I checked the name of my web-container. For my specific setup the name was "ddev-ddev-vscode-devcontainer-drupal9-template-web", then i turned this name into a hex value, using:
~~~
printf ddev-ddev-vscode-devcontainer-drupal9-template-web | od -A n -t x1 | tr -d '[\n\t ]'
~~~
after I had the hex-value i put it into my uri and added the path to my workspace (container-path) and got this solution:
~~~
code --folder-uri="vscode-remote://attached-container+646465762d646465762d7673636f64652d646576636f6e7461696e65722d64727570616c392d74656d706c6174652d776562%/var/www/html"
~~~
Now i can easily attach to my container and open the correct workspace using a code.sh in my project!

Only thing left is using my code.sh in ddevs post-start hooks and custom commands, but both are using the Container CLI which doesn't know "code".

**Note:**
I defined a "workspaceFolder" variable in my devcontainer.json:
~~~
"workspaceFolder": "/home/user/myprojects/ddev-vscode-devcontainer-drupal9-template"
~~~
I'm not sure if this is even needed nor if the devcontainer is used in general! VSCode's cache is super wierd, i need to do some more testing to find out.

I hope this helps! Here are some Resources i used:
https://stackoverflow.com/questions/60379221/how-to-attach-a-remote-container-using-vscode-command-line/63715551#63715551
https://github.com/geircode/vscode-attach-to-container-script

Have a nice day everyone :)

**Comment by Morgy93** [comment]
The solution sounds crazy, but glad you figured a way out.

I read that you're also using WSL (hopefully v2) so do you notice any performance difference between working on WSL vs. working inside the container?

I'm using multiple different projects with ddev and even high performance nodejs development (hot reload, auto bundle, etc.) with Docker for Windows + WSL2 and I don't notice any performance degradation. (Could it be any faster? )

The other thing would be that I have all my extensions running on WSL which works just fine, but with "attach to container" I would have to manage them all individually for each container. 

Would really love to contribute here, but currently I don't see any benefit (at least for me or using WSL2 in general).

**Comment by JPustkuchen** [comment]
Hi @Morgy93,

> The other thing would be that I have all my extensions running on WSL which works just fine, but with "attach to container" I would have to manage them all individually for each container.

Well yes I think it depends on your amount and type of projects. For us it's just the other way around. We want to predefine certain configuration types for different projects by DDEV / .vscode templates so that if you for example work in a Drupal Project today, have all Drupal best-practices and on another day work on Wordpress with a different tooling / configuration. If on a third day you work on a VUE application or your teammate needs to set up the Drupal Environments, you're happy you don't have to put everything combined on your (WSL2 / local) machine... that's the idea behind.

**Comment by rfay** [comment]
It's interesting that anybody would even take this on on WSL2, which has basically native performance, this and Linux would be the last place I'd see undertaking this. I'm sure you know that vscode has lovely integration with WSL2, just `code .` from the WSL2 project. https://ddevhq.org/ddev-local/ddev-wsl2-getting-started/

**Comment by Morgy93** [comment]
> It's interesting that anybody would even take this on on WSL2, which has basically native performance, this and Linux would be the last place I'd see undertaking this. I'm sure you know that vscode has lovely integration with WSL2, just `code .` from the WSL2 project. https://ddevhq.org/ddev-local/ddev-wsl2-getting-started/

Yes that's of course what I'm using, it's basically just the https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl extension.

I understand the needs by @JPustkuchen but I still think that's not the "optimal" way to go. (Not even regarding the crazy setup it needs upfront)

You can set up workspaces with VSCode which are awesome. (ref. https://code.visualstudio.com/docs/editor/workspaces)

So basically I just have multiple workspaces with lots of different settings for all the projects.
This allows me to have the "different tooling / configuration" as mentioned for every project. (Or every workspace, they are not limited to projects - even multi-root workspaces are supported!)
This file is even commited to git so that everyone in the team benefits from changes.

**Comment by JPustkuchen** [comment]
@rfay We're using WSL / code . for a long period of time now and it's truely awesome :)

@Morgy93: Well I think our approach is just 1 step "further" (different). You're sharing the VSCode configuration, we'd also like to share the configuration of the dev system behind. And I think that's even better as otherwise you'll need custom setup on your machine if shared in team.

Your approach is kind of developer-focused (one dev machine with all tools) while our plan is more project-focused, where each dev project brings its dev environment and vscode settings combined and ready to code

... if it works one day ;)

**Comment by JPustkuchen** [comment]
Anyway I'd be happy if someone likes to point out the pro's and con's of the solutions... for what we need (massive multi-project) I still think it makes sense? Or am I missing something completely?

**Comment by rfay** [comment]
On WSL2 there are so very many solutions these days. One new one is to use the almost-released feature where you can run any linux gui app inside wsl2. It's still an insider feature, but it works great, and solves all the ambiguity problems that WSL2 spends so much time dealing with. So you just run vscode or PHPStorm inside WSL2. works fine.

**Comment by JPustkuchen** [comment]
For all here, things are not perfect yet as some integration points are missing, but it works great already and allows to setup (for example Drupal 9) VSCode IDE with XDebug, PHPUnit, ... in seconds (the rest is waiting time) and without special configuration. Especially a great start for new team members or quick tests / developments in isolated containers.

Feel free to help: https://github.com/webksde/ddev-vscode-devcontainer-drupal9-template

Thanks to @joshsedl who did most of the fantastic work :)
Already more than just a proof of concept but lots of opportunities left. :)

**Comment by tyler36** [comment]
I'm interested too. I run my project from WSL on a Win10 system. 

I hit an issue with my phpunit workflow.

- Running phpunit on WSL works; when a test fails, it shows the full path and line number relative to my host OS. VSCode allows me to click through and open it

``` shell
phpunit
There was 1 failure:
/home/user13/code/laravel/tests/Unit/CommentModelTest.php:42
```

- When I run phpunit from inside the container, the failing test file path is inside the code so VSCode can't see the file.

``` shell
ddev phpunit
There was 1 failure:
/var/www/html/tests/Unit/CommentModelTest.php:42
```

Originally, I thought if I could open the web container directly with `Remote: container` extension, this would work well.
However, I wonder if a better solution might be to open the container with `Remote: SSH` . Unfortunately, I'm having an issue with this ATM. https://github.com/drud/ddev-contrib/issues/198

**Comment by tyler36** [comment]
@rfay I use VSCode

**Comment by rfay** [comment]
Sorry for the brain-fart response. Glad you're on track with the ssh approach. I would imagine that if you're using Docker Desktop you should also be able to get the docker-container approach to work.

**Comment by rfay** [comment]
As far as I can tell, this isn't something that requires anything from DDEV. However, we may want to add a bit to the docs if anybody is enjoying working this way.  There are SO MANY WAYS to do ALL THE THINGS! Closing for now.
